packages/glob/src/internal-pattern-helper.ts

```diff
@@ -0,0 +1,142 @@
+import * as pathHelper from './internal-path-helper'
+import {IGlobOptions} from './internal-glob-options'
+import {MatchResult} from './internal-match-result'
+import {Pattern} from './internal-pattern'
+
+const IS_WINDOWS = process.platform === 'win32'
+
+/**
+ * Given an array of patterns, returns an array of paths to search.
+ * Duplicates and paths under other included paths are filtered out.
+ */
+export function getSearchPaths(patterns: Pattern[]): string[] {
+  // Ignore comment and negate patterns
+  patterns = patterns.filter(x => !x.comment && !x.negate)
+
+  // Create a map of all search paths
+  const searchPathMap: {[key: string]: string} = {}
+  for (const pattern of patterns) {
+    const key = IS_WINDOWS
+      ? pattern.searchPath.toUpperCase()
+      : pattern.searchPath
+    searchPathMap[key] = 'candidate'
+  }
+
+  const result: string[] = []
+
+  for (const pattern of patterns) {
+    // Check if already included
+    const key = IS_WINDOWS
+      ? pattern.searchPath.toUpperCase()
+      : pattern.searchPath
+    if (searchPathMap[key] === 'included') {
+      continue
+    }
+
+    // Check for an ancestor search path
+    let foundAncestor = false
+    let tempKey = key
+    let parent = pathHelper.dirname(tempKey)
+    while (parent !== tempKey) {
+      if (searchPathMap[parent]) {
+        foundAncestor = true
+        break
+      }
+
+      tempKey = parent
+      parent = pathHelper.dirname(tempKey)
+    }
+
+    // Include the search pattern in the result
+    if (!foundAncestor) {
+      result.push(pattern.searchPath)
+      searchPathMap[key] = 'included'
+    }
+  }
+
+  return result
+}
+
+/**
+ * Matches the patterns against the path
+ */
+export function match(patterns: Pattern[], itemPath: string): MatchResult {
+  let result: MatchResult = MatchResult.None
+
+  for (const pattern of patterns) {
+    if (pattern.comment) {
+      continue
+    }
+
+    if (pattern.negate) {
+      result &= ~pattern.match(itemPath)
+    } else {
+      result |= pattern.match(itemPath)
+    }
+  }
+
+  return result
+}
+
+/**
+ * Parses the pattern strings into Pattern objects
+ */
+export function parse(patterns: string[], options: IGlobOptions): Pattern[] {
+  const result: Pattern[] = []
+
+  for (let patternString of patterns) {
+    // Parse
+    const pattern = new Pattern(patternString)
+
+    // Comment
+    if (pattern.comment) {
+      continue
+    }
+
+    // Push
+    result.push(pattern)
+
+    // Implicitly match descendant paths
+    if (options.implicitDescendants && !endsWithGlobstar(patternString)) {
```

Comment:
What if my `pattern` is `dir/*`?
It fails the `endsWithGlobstar` check and you'd end up with the pattern `dir/*/**`

Is that still a valid pattern and will it achieve the desired result? (Desired result being that users don't need to specify an exact globstar pattern to match directories)

Ideal behavior might be that `dir/*` is converted into `dir/**` for the user.

Reply:
`dir/*/**` and `dir/**` are not the same and (the format does not return `dir` and the latter does)

`dir/*/**` is the desired result

Reply:
What does the input `dir/*` mean then as a pattern?

Reply:
`dir/*` means all children of dir

Reply:
`shopt -s globstar` and `echo` is helpful to test different globs